allennlp/common/registrable.py

```diff
@@ -3,17 +3,20 @@
 any base class with a named registry for its subclasses and a decorator
 for registering them.
 """
-from collections import defaultdict
-from typing import TypeVar, Type, Callable, Dict, List, Optional, Tuple
 import importlib
 import logging
+from collections import defaultdict
+from typing import Callable, ClassVar, DefaultDict, Dict, List, Optional, Tuple, Type, TypeVar, cast
 
 from allennlp.common.checks import ConfigurationError
 from allennlp.common.from_params import FromParams
 
 logger = logging.getLogger(__name__)
 
-T = TypeVar("T", bound="Registrable")
+_T = TypeVar("_T")
+_RegistrableT = TypeVar("_RegistrableT", bound="Registrable")
+
+_SubclassRegistry = Dict[str, Tuple[type, Optional[str]]]
```

Comment:
Should this be:

```suggestion
_SubclassRegistry = Dict[str, Tuple[Type, Optional[str]]]
```

Reply:
It could be but it doesn't need to me. If I'm not mistaken, `type` is equivalent to `Type[object]` whereas `Type` is equivalent to `Type[Any]`. Using a bare `Type` means that type checkers will accept anything you want to do like it would annotating it with `Any`. By using `type`, we're saying that you can't make too many assumptions about the `type` you get back. This makes sense because you're using a `getattr(...)` to get the constructor.

This is also why I have to do the `cast(Type[_RegistrableT], subclass)` to be explicit about what we're returning. Some settings for type checkers will yell at you for that. This is like using `cast(...)` on something annotated as `object`.